File: ios/TerminalDeck/Ports/LocalhostAddress.swift
/**
 * What somebody typed into the Localhost tab's address field, turned into a port
 * on the machine and a path to ask it for — or into the sentence explaining why
 * it is not one.
 *
 * Asad, on the phone: *"the `+` starts a new browsing window here too"*, and
 * *"for localhost I want it exactly the way I explained for the Mac app — PC-to-PC
 * control. The same way I want all the controls here for iOS too."*
 *
 * ## Why a `+` was needed at all, when there is already a list
 *
 * Because the list can only ever offer what the machine reported, at the root of
 * the site. Three ordinary things were unreachable from it and each of them is
 * the sort of thing this feature exists for:
 *
 *  - **a path.** `localhost:3000/admin` is where the thing being worked on
 *    actually is, and every row on that screen opens `/`.
 *  - **a port that has just come up.** The list is a `ports` frame the desktop
 *    pushes; a server started twenty seconds ago is not on it until the next
 *    scan, and until then the phone can see it is missing but cannot ask for it.
 *  - **a port the grouping folded away.** Three of the five groups start closed,
 *    on purpose, and opening a group to find one number is not the gesture
 *    somebody who already knows the number wants.
 *
 * The tunnel underneath does not have any of those limits: `tunnel.open` on the
 * desktop is checked against a **fresh scan** rather than against whatever was
 * last offered, so a typed port either answers or is refused by the machine with
 * a sentence. See `src/main/remote/tunnel.ts`.
 *
 * ## A live link is **not** refused any more, and here is what changed
 *
 * This file used to end its argument with *"a live link is refused, and that
 * refusal is the answer"*: driving the desktop's own browser from a paired
 * device is blocked at every one of the five browser tools in
 * `src/main/deck-control/browser-tools.ts`, because a phone that can make
 * somebody's Mac open a page and raise a banner asking for a password — inside
 * their own trusted chrome — is a phishing primitive with the best possible
 * disguise.
 *
 * That is still true of the *tools*, and it was never true of the **machine's
 * own browser opened by its owner's own phone**. Asad, having typed a web
 * address and been told the machine could not:
 *
 * > *"browsers should browse any normal Google or any web internet website
 * > also. But it will be actually browsing on the server side; here it will be
 * > presenting that. So it shouldn't say that it cannot browse, because before I
 * > failed to browse. So this browser is not only for local, it is for internet
 * > also, for live websites also. So it should work seamless for everything."*
 *
 * **The refusal he hit was a missing wire, not a policy**, and the wire is
 * named: `CAPABILITY.web` is advertised only when `RemoteEndpointOptions.openUrl`
 * is a function — `src/main/remote/server.ts` line 2280, `if (name ===
 * CAPABILITY.web) return typeof options.openUrl === 'function'` — and the
 * headless host passed none. `src/headless/host.ts` says so in its own words
 * where it finally passes one: *"the refusal was not a policy — it was a missing
 * wire… `web.open` is backed by `openUrl`, this host passed none, and
 * `capabilitiesFor` therefore never advertised `web` — so the phone's address
 * bar took one look at `canOpenPages`, decided the machine could not do it, and
 * printed a sentence explaining that a site would load on the phone instead. The
 * sentence was true of a world where this option did not exist."*
 *
 * So the phone's job is no longer to refuse; it is to **decide which of three
 * things a typed line is**, and `classify` is that decision. `parse` below is
 * unchanged and is still the narrower question — *is this one of this machine's
 * own ports* — which is what the tunnel needs and what its own tests pin.
 *
 * ## What counts as the machine
 *
 * The loopback names, and only those: `localhost`, `127.0.0.1` (and the rest of
 * `127/8`, which is all loopback), and `::1` in both its bare and bracketed
 * spellings. A bare number is treated as a port on it, because somebody typing
 * `3000` into a field on the Localhost tab means the obvious thing.
 *
 * Deliberately **not** the machine's LAN address. The tunnel dials the machine's
 * *own* loopback, so an address that resolves somewhere else would be a request
 * this app cannot serve however plausible it looks — and one that quietly served
 * a different computer's page under the machine's name would be worse than a
 * refusal.
 */

import Foundation

enum LocalhostAddress {

    /// One typed line, resolved.
    enum Parsed: Equatable {
        /// A port on the machine, and the path to ask it for. The path always
        /// begins with `/` — `PortTunnel` hands back an origin and this is
        /// resolved against it, so a relative one would be dropped.
        case address(port: Int, path: String)
        /// Not something this screen can open, and the sentence to show. Written
        /// as a whole sentence rather than a code, because it is drawn under the
        /// field the moment it is true and nothing else will explain it.
        case refused(String)
    }

    /**
     * One typed line, as the app can act on it. Three outcomes are a *place to
     * go* and the fourth is a sentence.
     */
    enum Typed: Equatable {
        /// One of this machine's own ports — a tunnel, opened in this phone's
        /// own web view on a real loopback origin, so the page gets cookies, a
        /// service worker and the WebSocket a dev server's hot reload runs on.
        case tunnel(port: Int, path: String)
        /// Somewhere on the web, normalised to a URL the machine will accept.
        /// It opens **on the machine**, in the machine's own browser.
        case page(String)
        /// Not an address at all, so it is a search — with the words as typed,
        /// for the sentence that confirms it, and the URL that performs it.
        case search(query: String, url: String)
        /// It **is** an address and this app will not open it: a scheme that is
        /// not http or https, a port outside 1–65535, a paste with a control
        /// character in it. The sentence is written to be shown under the field.
        case refused(String)
    }

    /**
     * Where a search goes.
     *
     * One constant rather than a setting, because a setting nobody has asked for
     * is a screen to maintain and a preference to migrate. Every browser on this
     * phone has a default and this is that default; it is a `https` URL on a
     * host the machine resolves, so it goes down exactly the same path as any
     * other page typed here and needs no special case at either end.
     */
    static let searchBase = "https://www.google.com/search?q="

    /**
     * **The one decision the address bars make.** Which of three things is this?
     *
     * > *"So this browser is not only for local, it is for internet also, for
     * > live websites also. So it should work seamless for everything."*
     *
     * Seamless means there is no mode to choose, so this is where the choosing
     * happens — once, over a `String`, with no view and no capability in scope.
     * Both address fields in the app call it and neither has an opinion of its
     * own; what a caller decides is only *which door* a `page` goes through, and
     * `MachineBrowserView` argues that separately.
     *
     * ## The rules, in the order they are applied, and why each is where it is
     *
     *  1. **Whitespace inside means a search.** A URL cannot contain a space —
     *     it percent-encodes one — so `what is my ip` was never an address, and
     *     a browser that answered "that is not an address this phone can read"
     *     to a question is the flat refusal this whole change exists to delete.
     *  2. **A control character means a paste went wrong**, or is smuggling a
     *     second target past the eye. Refused rather than searched, because the
     *     honest reading is that the input is damaged. `browser-url.ts` makes
     *     the same check on the host for the same reason.
     *  3. **A bare number is a port.** `3000` typed into this app means the
     *     obvious thing, and `:3000` is the same habit with a colon. An
     *     out-of-range number is refused rather than searched: somebody who
     *     typed `70000` meant a port and wants to know it is not one.
     *  4. **A scheme that is not http(s) is refused** — `file:`, `ws:`,
     *     `javascript:` — and the check has to come before anything reads the
     *     string as a host, because otherwise `file:///etc/passwd` becomes the
     *     host `file` and is opened as a page. It also has to *not* fire on
     *     `localhost:3000`, whose "scheme" is `localhost`: that is the single
     *     most likely thing anybody types here and `new URL` gets it wrong.
     *     `hostAndPort` is the same shape `browser-url.ts` matches for the same
     *     trap, kept in step deliberately.
     *  5. **This machine's own loopback is the tunnel**, and `parse` owns that
     *     answer including its refusal for a loopback name with no port —
     *     *"Which port?"* is a question with an answer somebody can type, and a
     *     search for the word `localhost` is not.
     *  6. **Anything that looks like a host is a page**; anything else is a
     *     search. `looksLikeAHost` is the whole of that judgement.
     *
     * ## Why `google.com` is the case this is measured against
     *
     * Nobody types the scheme. `web.open` on the host runs what it is given
     * through `normalizeUrl` in `browser-url.ts`, which completes a bare host
     * itself — but `isNavigationAllowed`, the guard on a *page-initiated*
     * navigation, does not, and `new URL('google.com')` throws. Completing it
     * here means what goes on the wire is a thing every gate on the far side
     * accepts, rather than a string whose fate depends on which gate reads it.
     */
    static func classify(_ raw: String) -> Typed {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else {
            return .refused("Type an address, a port on this machine, or something to search for.")
        }

        if text.contains(where: \.isWhitespace) { return searching(text) }
        if text.unicodeScalars.contains(where: { $0.value < 0x20 || $0.value == 0x7f }) {
            return .refused("That contains characters an address cannot contain.")
        }

        // A bare port, with or without the colon somebody types out of habit.
        // Handed to `parse` rather than parsed again here, so the range check
        // and its sentence exist once.
        let bare = text.hasPrefix(":") ? String(text.dropFirst()) : text
        if bare.allSatisfy(\.isNumber) { return fromParse(text) }

        if let scheme = explicitScheme(text), scheme != "http", scheme != "https" {
            return .refused("Only http and https pages can be opened, not \(scheme):.")
        }

        // Protocol-relative — `//example.com` — is given a scheme rather than a
        // second pair of slashes. Vanishingly rare from a thumb and handled all
        // the same, because `browser-url.ts` handles it on the host and a string
        // the two ends disagree about is a refusal nobody can explain.
        let withScheme = text.contains("://")
            ? text
            : (text.hasPrefix("//") ? "http:\(text)" : "http://\(text)")
        guard var parts = URLComponents(string: withScheme),
              let host = parts.host, !host.isEmpty else {
            // Not readable as a URL at all — `a b`, `??`, a stray bracket. A
            // browser searches for it rather than lecturing about grammar.
            return searching(text)
        }

        if isLoopback(host) { return fromParse(text) }

        guard looksLikeAHost(host, hasPort: parts.port != nil) else { return searching(text) }

        // Lowercased on the way out as well as on the way through the checks, so
        // `HTTP://GOOGLE.COM` and `google.com` reach the machine as one string.
        parts.scheme = parts.scheme?.lowercased()
        parts.host = host.lowercased()
        return .page(parts.string ?? withScheme)
    }

    /// The tunnel half, borrowed from `parse` so that its answers and its
    /// sentences exist in exactly one place.
    private static func fromParse(_ text: String) -> Typed {
        switch parse(text) {
        case let .address(port, path): return .tunnel(port: port, path: path)
        case let .refused(why): return .refused(why)
        }
    }

    /// Words, and the address that searches for them. Percent-encoded against
    /// `alphanumerics` rather than a query character set: it over-encodes, and
    /// over-encoding a query is always safe while under-encoding one silently
    /// changes what was searched for.
    private static func searching(_ query: String) -> Typed {
        let encoded = query.addingPercentEncoding(withAllowedCharacters: .alphanumerics) ?? query
        return .search(query: query, url: searchBase + encoded)
    }

    /**
     * `host:port` with an optional path — the shape that fools `new URL` and
     * `URL(string:)` alike into reading the host as a scheme.
     *
     * The same expression `browser-url.ts` matches on the host, deliberately: a
     * string this side reads as a host and that side reads as a scheme is a
     * refusal nobody can explain from either end.
     */
    private static let hostAndPort = try? NSRegularExpression(
        pattern: "^(?:[A-Za-z0-9-]+(?:\\.[A-Za-z0-9-]+)*|\\[[0-9A-Fa-f:]+\\]):[0-9]{1,5}(?:[/?#].*)?$")

    /// The scheme a line actually has, or nil — including nil for the two shapes
    /// that only look like they have one.
    private static func explicitScheme(_ text: String) -> String? {
        let whole = NSRange(text.startIndex ..< text.endIndex, in: text)
        if hostAndPort?.firstMatch(in: text, range: whole) != nil { return nil }
        guard let colon = text.firstIndex(of: ":") else { return nil }
        let scheme = String(text[text.startIndex ..< colon])
        guard let first = scheme.first, first.isLetter,
              scheme.allSatisfy({ $0.isLetter || $0.isNumber || $0 == "+" || $0 == "." || $0 == "-" })
        else { return nil }
        return scheme.lowercased()
    }

    /**
     * Whether a host name is a name somebody meant, rather than a word.
     *
     * Four ways to be one, and the last is the interesting one:
     *
     *  - it carries a **port**, which nobody writes on a search term;
     *  - it is an **IPv6 literal**, which arrives bracketed;
     *  - it is an **IPv4 literal**;
     *  - it has a **dot** and its last label is two or more letters. That is the
     *    rule every browser uses and it is deliberately not a list of real
     *    top-level domains: the list changes, a phone that shipped last year
     *    would refuse a domain that exists, and being wrong about `.zip` costs
     *    nothing next to being wrong about a domain a person is looking at.
     *
     * `git`, `readme`, `terminal deck` and `1.2.3` are all *not* hosts, which is
     * what makes them searches.
     */
    private static func looksLikeAHost(_ host: String, hasPort: Bool) -> Bool {
        if hasPort { return true }
        let name = host.lowercased()
        if name.hasPrefix("[") { return true }
        if isIPv4(name) { return true }
        let labels = name.split(separator: ".", omittingEmptySubsequences: false)
        guard labels.count >= 2, labels.allSatisfy({ !$0.isEmpty }) else { return false }
        guard let last = labels.last, last.count >= 2, last.allSatisfy(\.isLetter) else {
            return false
        }
        return true
    }

    /// A dotted quad, each octet in range. Shared by `looksLikeAHost` and
    /// `isLoopback`, which asks the same question of the same shape.
    private static func isIPv4(_ name: String) -> Bool {
        let parts = name.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 4, parts.allSatisfy({ !$0.isEmpty && $0.allSatisfy(\.isNumber) }) else {
            return false
        }
        return parts.compactMap { Int($0) }.allSatisfy { $0 >= 0 && $0 <= 255 }
    }

    /**
     * Turn a line into a port and a path.
     *
     * Written as a static function over a `String` rather than as a
     * `URLComponents` extension because two of the four shapes people type are
     * not URLs at all — `3000` and `localhost:3000` both parse as a *path* under
     * `URLComponents`, with `localhost` as a scheme-less path segment and `3000`
     * as, remarkably, nothing at all. Handling those first and only then handing
     * the rest to `URL` is what makes the common case work.
     */
    static func parse(_ raw: String) -> Parsed {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else {
            return .refused("Type a port, or an address on this machine.")
        }

        // A bare port, with or without the colon somebody types out of habit.
        let bare = text.hasPrefix(":") ? String(text.dropFirst()) : text
        if bare.allSatisfy(\.isNumber) {
            return port(Int(bare) ?? 0, path: "/")
        }

        /*
         * Everything else is a URL, and one is given a scheme if it has none:
         * `URL(string:)` accepts `localhost:3000/admin` and reads `localhost` as
         * the *scheme*, which puts the port in the path and the host nowhere.
         *
         * `URLComponents` rather than `URL` for the reading, and the difference
         * is a trailing slash. `URL.path` normalises one away — `/a/b/` comes
         * back as `/a/b` — and that is not cosmetic on a dev server: a directory
         * URL and a file URL resolve relative assets from different places, so
         * the page would load and its stylesheet would 404. The percent-encoded
         * forms are used for the same class of reason: a path with a space or a
         * `#` in it survives the round trip instead of being re-encoded wrongly
         * when it is resolved against the tunnel's origin.
         */
        let withScheme = text.contains("://") ? text : "http://\(text)"
        guard let parts = URLComponents(string: withScheme), let host = parts.host else {
            return .refused("That is not an address this phone can read.")
        }

        guard let scheme = parts.scheme?.lowercased(), scheme == "http" || scheme == "https" else {
            return .refused("Only http and https pages can be opened here.")
        }

        guard isLoopback(host) else {
            /*
             * A fact about **this function**, and nothing more.
             *
             * `parse` answers one question — *is this one of this machine's own
             * ports* — so a host that is not loopback is a no here and is not a
             * refusal anywhere a person can see. `classify` is what the address
             * bars call, and it takes this case and opens it on the machine.
             * Nothing draws this string; it is kept because `parse` is a public
             * answer with its own tests and *"no"* with no reason attached is a
             * worse value to hand back than one with a sentence on it.
             */
            return .refused("\(host) is not a page this machine is serving.")
        }

        guard let number = parts.port else {
            return .refused("Which port? A page on this machine is reached by its port number.")
        }

        var path = parts.percentEncodedPath.isEmpty ? "/" : parts.percentEncodedPath
        if let query = parts.percentEncodedQuery, !query.isEmpty { path += "?\(query)" }
        if let fragment = parts.percentEncodedFragment, !fragment.isEmpty { path += "#\(fragment)" }
        return port(number, path: path)
    }

    /// The range check, in one place, so a bare number and a full URL are refused
    /// with the same sentence. Port 0 is excluded as well as the out-of-range
    /// values: nothing listens on it, and `PortTunnel` refuses it a step later
    /// with a message about this phone rather than about what was typed.
    private static func port(_ number: Int, path: String) -> Parsed {
        guard number > 0, number <= 65_535 else {
            return .refused("A port is a number between 1 and 65535.")
        }
        return .address(port: number, path: path)
    }

    /**
     * Whether a host name means *this machine's own loopback*.
     *
     * `127.0.0.0/8` in full rather than the one address people type, because a
     * dev server bound to `127.0.0.2` is a real thing — it is how two projects
     * share a port — and the desktop's own scan reports whatever it finds on the
     * loopback interface rather than only the canonical address.
     */
    private static func isLoopback(_ host: String) -> Bool {
        let name = host.lowercased()
        if name == "localhost" || name == "::1" || name == "[::1]" { return true }
        guard isIPv4(name) else { return false }
        return name.split(separator: ".").first.flatMap { Int($0) } == 127
    }
}

File: android/app/src/main/java/dev/terminaldeck/android/LocalhostAddress.kt
package dev.terminaldeck.android

import java.net.URI
import java.util.Locale

/**
 * What somebody typed into a browser address field, turned into a place to go — a port on the
 * machine, a page on the machine, a search — or into the sentence explaining why it is none of those.
 *
 * A 1:1 transcription of `ios/TerminalDeck/Ports/LocalhostAddress.swift`. Android's Localhost tab
 * never had a typed address field — it only opened ports by number — so this pure decision did not
 * exist on this side and is added here rather than in the `protocol` package, mirroring the iOS
 * `Ports/` home. Both the machine-browser address bars and the in-session floating window call it, so
 * `google.com`, `https://…`, `3000` and `what is my ip` mean the same thing everywhere.
 *
 * Asad, having typed a web address and been told the machine could not browse it:
 *
 *   > *"browsers should browse any normal Google or any web internet website also. But it will be
 *   > actually browsing on the server side; here it will be presenting that… So this browser is not
 *   > only for local, it is for internet also, for live websites also. So it should work seamless for
 *   > everything."*
 *
 * Seamless means there is no mode to choose, so [classify] is where the choosing happens — once, over
 * a `String`, with no view and no capability in scope. What a caller decides is only *which door* a
 * [Typed.Page] goes through: a tunnel opens on this phone, a page opens on the machine.
 */
object LocalhostAddress {

    /** One typed line, as the app can act on it. Three outcomes are a place to go; the fourth is a
     *  sentence written to be shown under the field. */
    sealed interface Typed {
        /** One of this machine's own ports — opened over a tunnel. */
        data class Tunnel(val port: Int, val path: String) : Typed

        /** Somewhere on the web, normalised to a URL the machine will accept — opens on the machine. */
        data class Page(val url: String) : Typed

        /** Not an address, so a search: the words as typed (for the sentence that confirms it) and the
         *  URL that performs it. */
        data class Search(val query: String, val url: String) : Typed

        /** It is an address this app will not open — a non-http(s) scheme, a port out of range, a
         *  control character. The sentence is drawn under the field. */
        data class Refused(val why: String) : Typed
    }

    /** [parse]'s narrower answer — *is this one of this machine's own ports* — kept because the
     *  tunnel needs it and its refusals name a real reason. */
    sealed interface Parsed {
        data class Address(val port: Int, val path: String) : Parsed
        data class Refused(val why: String) : Parsed
    }

    /**
     * Where a search goes. One constant rather than a setting, because a setting nobody has asked for
     * is a screen to maintain — every browser has a default and this is that default. It is an https
     * URL on a host the machine resolves, so it goes down the same path as any other typed page.
     */
    const val SEARCH_BASE = "https://www.google.com/search?q="

    /**
     * **The one decision the address bars make.** Which of three things is this?
     *
     * The rules, in the order they are applied (each argued at length in the iOS source):
     *
     *  1. Whitespace inside means a search — a URL percent-encodes a space, so `what is my ip` was
     *     never an address, and refusing to read a question is the flat refusal this exists to delete.
     *  2. A control character means a paste went wrong; refused rather than searched.
     *  3. A bare number is a port; `:3000` is the same habit with a colon. Out of range is refused.
     *  4. A scheme that is not http(s) is refused — and the check comes *before* anything reads the
     *     string as a host, or `file:///etc/passwd` becomes the host `file`. It must not fire on
     *     `localhost:3000`, whose "scheme" is `localhost` — [hostAndPort] is that guard.
     *  5. This machine's own loopback is the tunnel; [parse] owns that answer and its refusals.
     *  6. Anything that looks like a host is a page; anything else is a search.
     */
    fun classify(raw: String): Typed {
        val text = raw.trim()
        if (text.isEmpty()) {
            return Typed.Refused("Type an address, a port on this machine, or something to search for.")
        }

        if (text.any { it.isWhitespace() }) return searching(text)
        if (text.any { it.code < 0x20 || it.code == 0x7f }) {
            return Typed.Refused("That contains characters an address cannot contain.")
        }

        // A bare port, with or without the colon somebody types out of habit — handed to `parse` so
        // the range check and its sentence exist once.
        val bare = if (text.startsWith(":")) text.drop(1) else text
        if (bare.isNotEmpty() && bare.all { it.isDigit() }) return fromParse(text)

        val scheme = explicitScheme(text)
        if (scheme != null && scheme != "http" && scheme != "https") {
            return Typed.Refused("Only http and https pages can be opened, not $scheme:.")
        }

        // Protocol-relative — `//example.com` — is given a scheme rather than a second pair of
        // slashes, matching the host's own `browser-url.ts`.
        val withScheme = when {
            text.contains("://") -> text
            text.startsWith("//") -> "http:$text"
            else -> "http://$text"
        }
        val uri = readUri(withScheme)
        val host = uri?.host
        if (uri == null || host.isNullOrEmpty()) {
            // Not readable as a URL at all — `a b`, `??`, a stray bracket. A browser searches for it
            // rather than lecturing about grammar.
            return searching(text)
        }

        if (isLoopback(host)) return fromParse(text)

        if (!looksLikeAHost(host, hasPort = uri.port != -1)) return searching(text)

        // Lowercased on the way out so `HTTP://GOOGLE.COM` and `google.com` reach the machine as one
        // string. The raw path/query/fragment are carried through untouched — re-encoding a path with
        // a space or a `#` in it silently changes what was asked for.
        val outScheme = (uri.scheme ?: "http").lowercase(Locale.ROOT)
        val out = StringBuilder().append(outScheme).append("://")
        uri.rawUserInfo?.let { out.append(it).append('@') }
        out.append(host.lowercase(Locale.ROOT))
        if (uri.port != -1) out.append(':').append(uri.port)
        out.append(uri.rawPath.orEmpty())
        uri.rawQuery?.let { out.append('?').append(it) }
        uri.rawFragment?.let { out.append('#').append(it) }
        return Typed.Page(out.toString())
    }

    /** The tunnel half, borrowed from [parse] so its answers and sentences exist in one place. */
    private fun fromParse(text: String): Typed = when (val parsed = parse(text)) {
        is Parsed.Address -> Typed.Tunnel(parsed.port, parsed.path)
        is Parsed.Refused -> Typed.Refused(parsed.why)
    }

    /** Words, and the address that searches for them. Encoded to bytes and passing only ASCII
     *  alphanumerics through — it over-encodes, which is always safe, while under-encoding a query
     *  silently changes what was searched for. */
    private fun searching(query: String): Typed {
        val encoded = StringBuilder()
        for (byte in query.toByteArray(Charsets.UTF_8)) {
            val value = byte.toInt() and 0xff
            val ch = value.toChar()
            if (ch in 'A'..'Z' || ch in 'a'..'z' || ch in '0'..'9') {
                encoded.append(ch)
            } else {
                encoded.append('%').append(value.toString(16).uppercase(Locale.ROOT).padStart(2, '0'))
            }
        }
        return Typed.Search(query, SEARCH_BASE + encoded)
    }

    /** `host:port` with an optional path — the shape that fools a URL parser into reading the host as
     *  a scheme. The same expression the host's `browser-url.ts` matches, kept in step deliberately. */
    private val hostAndPort = Regex(
        "^(?:[A-Za-z0-9-]+(?:\\.[A-Za-z0-9-]+)*|\\[[0-9A-Fa-f:]+\\]):[0-9]{1,5}(?:[/?#].*)?$"
    )

    /** The scheme a line actually has, or null — including null for the two shapes that only look
     *  like they have one (`localhost:3000`, `127.0.0.1:8080`). */
    private fun explicitScheme(text: String): String? {
        if (hostAndPort.matches(text)) return null
        val colon = text.indexOf(':')
        if (colon < 0) return null
        val scheme = text.substring(0, colon)
        val first = scheme.firstOrNull() ?: return null
        if (!first.isLetter()) return null
        if (!scheme.all { it.isLetter() || it.isDigit() || it == '+' || it == '.' || it == '-' }) return null
        return scheme.lowercase(Locale.ROOT)
    }

    /**
     * Whether a host name is a name somebody meant, rather than a word. Four ways to be one, and the
     * last is the interesting one: a dot with a last label of two or more letters. That is the rule
     * every browser uses and it is deliberately not a list of real top-level domains — the list
     * changes, and being wrong about `.zip` costs nothing next to being wrong about a domain a person
     * is looking at. `git`, `readme` and `1.2.3` are all *not* hosts, which makes them searches.
     */
    private fun looksLikeAHost(host: String, hasPort: Boolean): Boolean {
        if (hasPort) return true
        val name = host.lowercase(Locale.ROOT)
        if (name.startsWith("[")) return true
        if (isIPv4(name)) return true
        val labels = name.split(".")
        if (labels.size < 2 || labels.any { it.isEmpty() }) return false
        val last = labels.last()
        return last.length >= 2 && last.all { it.isLetter() }
    }

    /** A dotted quad, each octet in range. */
    private fun isIPv4(name: String): Boolean {
        val parts = name.split(".")
        if (parts.size != 4 || parts.any { it.isEmpty() || !it.all { c -> c.isDigit() } }) return false
        return parts.all { (it.toIntOrNull() ?: -1) in 0..255 }
    }

    /**
     * Turn a line into a port and a path — the narrower question the tunnel needs. Two of the shapes
     * people type (`3000`, `localhost:3000`) are not URLs at all, so those are handled first and only
     * the rest is handed to a URL parser.
     */
    fun parse(raw: String): Parsed {
        val text = raw.trim()
        if (text.isEmpty()) return Parsed.Refused("Type a port, or an address on this machine.")

        val bare = if (text.startsWith(":")) text.drop(1) else text
        if (bare.isNotEmpty() && bare.all { it.isDigit() }) return port(bare.toIntOrNull() ?: 0, "/")

        val withScheme = if (text.contains("://")) text else "http://$text"
        val uri = readUri(withScheme)
        val host = uri?.host
        if (uri == null || host.isNullOrEmpty()) return Parsed.Refused("That is not an address this phone can read.")

        val scheme = uri.scheme?.lowercase(Locale.ROOT)
        if (scheme != "http" && scheme != "https") {
            return Parsed.Refused("Only http and https pages can be opened here.")
        }

        if (!isLoopback(host)) {
            // A fact about this function and nothing more: a host that is not loopback is a no here.
            // `classify` intercepts this case and opens it on the machine, so nothing draws this
            // string — but a public answer with a reason on it is worth more than a bare no.
            return Parsed.Refused("$host is not a page this machine is serving.")
        }

        if (uri.port == -1) {
            return Parsed.Refused("Which port? A page on this machine is reached by its port number.")
        }

        var path = uri.rawPath?.ifEmpty { "/" } ?: "/"
        uri.rawQuery?.let { if (it.isNotEmpty()) path += "?$it" }
        uri.rawFragment?.let { if (it.isNotEmpty()) path += "#$it" }
        return port(uri.port, path)
    }

    /** The range check in one place, so a bare number and a full URL are refused with the same
     *  sentence. Port 0 is excluded as well as the out-of-range values — nothing listens on it. */
    private fun port(number: Int, path: String): Parsed =
        if (number in 1..65_535) Parsed.Address(number, path)
        else Parsed.Refused("A port is a number between 1 and 65535.")

    /**
     * Whether a host name means this machine's own loopback. `127.0.0.0/8` in full rather than the one
     * address people type, because a dev server bound to `127.0.0.2` is a real thing.
     */
    private fun isLoopback(host: String): Boolean {
        val name = host.lowercase(Locale.ROOT)
        if (name == "localhost" || name == "::1" || name == "[::1]") return true
        if (!isIPv4(name)) return false
        return name.substringBefore(".").toIntOrNull() == 127
    }

    /** Read a URL string without throwing — a line that will not parse is a search, not a crash. */
    private fun readUri(text: String): URI? = try {
        URI(text)
    } catch (e: Exception) {
        null
    }
}
